File: 46_Permutations.py
# 46. Permutations

class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        self.dfs(nums, [], res)
        return res

    def dfs(self,nums, path,res):
        if not nums:
            res.append(path)
            return
        for i in range(len(nums)):
            self.dfs(nums[:i]+nums[i+1:],path+[nums[i]],res)

# 47. Permutations II

class Solution(object):
    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        res = []
        self.permutation(nums,[],res)
        return res


    def permutation(self,nums,path,res):
        if len(nums) == 0:
            if path not in res:
                res.append(path)
            return
        for i in range(len(nums)):
            self.permutation(nums[:i]+nums[i+1:],path+[nums[i]],res)

File: C++/46_Permutations.cpp
class Solution {
public:
    void dfs(vector<int> nums, vector<int>& path, vector<vector<int>>& result){
        if (nums.empty()){
            result.push_back(path);
            return;
        }

        for (int i=0;i<nums.size();i++){
            path.push_back(nums[i]);
            nums.erase(nums.begin()+i);
            dfs(nums, path, result);
            nums.insert(nums.begin()+i,path.back());
            path.pop_back();
        }
    }

    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> result = {};
        vector<int> path = {};
        dfs(nums, path, result);
        return result;
    }
};


class Solution {
public:
    void dfs(vector<int> nums, vector<int>& path, vector<vector<int>>& result){
        if (path.size() == nums.size()){
            result.push_back(path);
            return;
        }

        for (int i=0;i<nums.size();i++){
            if (find(path.begin(),path.end(),nums[i])==path.end()){
                path.push_back(nums[i]);
                dfs(nums, path, result);
                path.pop_back();
            }
        }
    }

    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> result = {};
        vector<int> path = {};
        dfs(nums, path, result);
        return result;
    }
};
